app/src/main/java/edu/outside2154/gamesense/model/Player.kt

```diff
@@ -1,31 +1,89 @@
 package edu.outside2154.gamesense.model
 
+import android.util.Log
+import com.google.firebase.database.DataSnapshot
+import com.google.firebase.database.DatabaseError
+import com.google.firebase.database.FirebaseDatabase
+import com.google.firebase.database.ValueEventListener
+import edu.outside2154.gamesense.model.Stat
+import edu.outside2154.gamesense.model.Boss
+import java.io.Serializable
 import java.util.*
 
 const val PLAYER_BASE_HEALTH = 100.0
 const val PLAYER_BASE_ATTACK = 100.0
 const val PLAYER_CRIT_MULT = 2.0
 
 
-class Player(_intStat: Stat, _atkStat: Stat, _regenStat: Stat) {
+class Player(androidId : String) : Serializable {
     var health = PLAYER_BASE_HEALTH
         private set
-    var intStat = _intStat
-        private set
-    var atkStat = _atkStat
-        private set
-    var regenStat = _regenStat
-        private set
+
+    var intStat : Stat? = null
+    var atkStat : Stat? = null
+    var regenStat : Stat? = null
+
     var currency = 0.0
         private set
     val pureDamage
-        get() = PLAYER_BASE_ATTACK * (atkStat.calcStat() ?: 0.0)
+        get() = PLAYER_BASE_ATTACK * (atkStat?.calcStat() ?: 0.0)
     val dead
         get() = health == 0.0
     private val rand = Random()
 
+    init {
+        // Create listener for data reading from Firebase
+        val fbListener = object : ValueEventListener {
```

Comment:
Our Firebase code is super boilerplate-y. We can probably [delegate the properties](https://kotlinlang.org/docs/reference/delegated-properties.html) instead to be cleaner. I can imagine something like this:

```kotlin
class FirebaseProperty<in R, T>(val root: Query, val transform: (DataSnapshot) -> T)
        : ReadWriteProperty<in R, T> {
    var _value: T

    init {
        root.addListenerForSingleValueEvent(object : ValueEventListener {
            override fun onDataChange(snapshot: DataSnapshot) {
                _value = transform(snapshot)
            }
            override fun onCancelled(databaseError: DatabaseError) {
                println("loadPost:onCancelled ${databaseError.toException()}")
            }
        })
    }

    operator fun getValue(thisRef: R, property: KProperty<*>): T = _value
    operator fun setValue(thisRef: R, property: KProperty<*>, value: T) { _value = value }
}

// Use as...
var atkStat: Stat by FirebaseProperty(atkFirebaseRef) { s ->
    Stat(s.child("goals").getValue(), s.child("current").getValue())
}
// ...which can probably be further abstracted to something like...
var atkStat: Stat by FirebaseStat(atkFirebaseRef)
```

I didn't check if this works (it probably doesn't), but the idea should.